File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofDesign/Components/PixelGlyphField.swift
import SwiftUI

/// An observable command channel for spawning tap-style ripples in a
/// `PixelGlyphField` from OUTSIDE the field. The field installs its own tap
/// gesture for ambient ripples; this lets a foreground view inject one
/// programmatically — e.g. tapping the agent avatar ripples the background
/// pixels behind it (an easter egg on the agent detail hero).
///
/// `point` is in the field's OWN coordinate space. A screen-anchored field
/// (`PixelWaveBackground(screenAnchored:)`) draws at global (0,0) sized to the
/// full screen, so a GLOBAL screen point maps 1:1 — pass the tapped view's
/// global center (`frame(in: .global)` midpoint).
@MainActor
@Observable
public final class GlyphRippleEmitter {
    public struct Pulse: Equatable {
        public var point: CGPoint
        /// Monotonic id so two ripples at the same point still read as distinct
        /// events (a plain point wouldn't trip `.onChange` on a repeat tap).
        public var token: Int
    }

    public private(set) var pulse = Pulse(point: .zero, token: 0)

    public init() {}

    /// Spawn a ripple centered on `point` (field-local / screen-anchored-global).
    public func emit(at point: CGPoint) {
        pulse = Pulse(point: point, token: pulse.token + 1)
    }
}

/// A stepped, organic field of small pixel "glyphs" that bloom and dissipate
/// like bacteria colonies or cloud poofs.
///
/// How it differs from `PixelDotBackground`: that engine is a *stateless* field
/// — each frame is a pure function of time, so it drifts smoothly forever. A
/// spreading/decaying colony instead needs *state that evolves in discrete
/// steps*. This view runs a small reaction-diffusion cellular automaton over the
/// dot grid. Designed seed "glyphs" are periodically stamped onto the grid; the
/// automaton grows them outward a little, hollows out saturated cores, and
/// erodes them with ambient decay — so each glyph stays small and focused, then
/// poofs away.
///
/// Cadence: the step interval cycles through `intervals` (a single, consistent
/// 150ms step by default; pass multiple values for an irregular heartbeat). The
/// display eases previous→current across the full gap to the next step, so
/// motion is continuous and smooth — no hold or pause between refreshes.
///
/// Only lit cells are drawn — the empty grid is never rendered, so the field
/// reads as a few scattered clusters on black rather than a full dot matrix.
public struct PixelGlyphField: View {
    private let intervals: [TimeInterval]
    private let baseColor: Color
    private let accentColor: Color?
    private let spacing: CGFloat
    private let dotSize: CGFloat
    private let peakOpacity: Double
    /// Optional external trigger so another view can ripple this field (see
    /// `GlyphRippleEmitter`). Nil for fields that only ripple on direct taps.
    private let rippleEmitter: GlyphRippleEmitter?

    @State private var sim = GlyphAutomaton()
    /// Reference time of the most recent simulation step — drives the brief
    /// previous→current "pop" so each tick eases in.
    @State private var lastStep: Date = .distantPast
    /// Duration of the gap until the next step. The crossfade spans this whole
    /// window so motion stays continuous — no hold or pause between steps.
    @State private var stepGap: TimeInterval = 0.3
    /// Active tap ripples — transient expanding rings of brightness layered on
    /// top of the automaton when the user taps the field.
    @State private var ripples: [Ripple] = []

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private struct Ripple {
        let origin: CGPoint
        let start: Date
    }

    // Tap-ripple tuning.
    private let rippleDuration: TimeInterval = 1.5
    private let rippleSpeed: Double = 430          // ring expansion, points / second
    private let rippleRingWidth: Double = 26       // ring softness, points
    private let ripplePeakOpacity: Double = 0.72   // ripple cells out-shine ambient

    /// `intervals` is the cadence pattern, cycled in order. A single value (the
    /// default 0.15s) gives a steady beat; multiple values vary it.
    public init(
        intervals: [TimeInterval] = [0.15],
        baseColor: Color = .white,
        accentColor: Color? = .appPrimary,
        spacing: CGFloat = 26,
        dotSize: CGFloat = 5.5,
        peakOpacity: Double = 0.45,
        rippleEmitter: GlyphRippleEmitter? = nil
    ) {
        self.intervals = intervals.isEmpty ? [0.3] : intervals
        self.baseColor = baseColor
        self.accentColor = accentColor
        self.spacing = spacing
        self.dotSize = dotSize
        self.peakOpacity = peakOpacity
        self.rippleEmitter = rippleEmitter
    }

    public var body: some View {
        let frozen = reduceMotion

        GeometryReader { proxy in
            TimelineView(.animation(paused: frozen)) { timeline in
                Canvas(opaque: false, rendersAsynchronously: false) { context, size in
                    draw(into: &context, size: size, now: timeline.date)
                }
            }
            // Build (or rebuild) the grid whenever the available size changes —
            // `initial: true` seeds it on first appear.
            .onChange(of: proxy.size, initial: true) { _, newSize in
                let cols = max(2, Int((newSize.width / spacing).rounded()))
                let rows = max(2, Int((newSize.height / spacing).rounded()))
                sim.configure(cols: cols, rows: rows)
                lastStep = .now
            }
            // Stepping loop. Alternates slow/fast intervals for the irregular
            // cadence. Restarts if Reduce Motion toggles (id: frozen); when
            // frozen it returns immediately and the field holds its last state.
            .task(id: frozen) {
                guard !frozen else { return }
                var idx = 0
                while !Task.isCancelled {
                    try? await Task.sleep(for: .seconds(intervals[idx]))
                    if Task.isCancelled { return }
                    sim.step()
                    lastStep = .now
                    idx = (idx + 1) % intervals.count
                    // Crossfade window = time until the NEXT step, so the ease
                    // fills the whole gap and the field never sits still.
                    stepGap = intervals[idx]
                }
            }
        }
        // Capture taps anywhere in the field to spawn a ripple. The field sits
        // behind the foreground content, so buttons/links still receive their
        // taps first; only taps that fall through to empty areas ripple.
        .contentShape(Rectangle())
        .onTapGesture { location in
            addRipple(at: location)
        }
        // External ripples (e.g. tapping the agent avatar). Bypasses hit testing,
        // so it still fires when the field itself is inert (`allowsHitTesting`
        // off behind a scroll). The pulse point is in this field's space.
        .onChange(of: rippleEmitter?.pulse) { _, newPulse in
            if let p = newPulse, p.token > 0 { addRipple(at: p.point) }
        }
    }

    /// Spawn a ripple at `location`, pruning any that have already faded out.
    private func addRipple(at location: CGPoint) {
        guard !reduceMotion else { return }
        let cutoff = Date(timeIntervalSinceNow: -rippleDuration)
        var active = ripples.filter { $0.start > cutoff }
        active.append(Ripple(origin: location, start: .now))
        if active.count > 8 { active.removeFirst(active.count - 8) }
        ripples = active
    }

    /// Combined brightness (0...1) from all active ripples for a cell at
    /// (px, py): each ripple is a soft expanding ring that fades with age.
    private func rippleLevel(px: CGFloat, py: CGFloat, now: Date) -> Double {
        if ripples.isEmpty { return 0 }
        var level = 0.0
        for ripple in ripples {
            let age = ripple.start.distance(to: now)
            if age < 0 || age > rippleDuration { continue }
            let radius = age * rippleSpeed
            let dx = Double(px - ripple.origin.x)
            let dy = Double(py - ripple.origin.y)
            let delta = (dx * dx + dy * dy).squareRoot() - radius
            let ring = exp(-(delta * delta) / (2 * rippleRingWidth * rippleRingWidth))
            let fade = 1 - age / rippleDuration
            level = max(level, ring * fade)
        }
        return level
    }

    private func draw(into context: inout GraphicsContext, size: CGSize, now: Date) {
        guard size.width > 0, size.height > 0, sim.cols > 0, sim.rows > 0 else { return }

        let cols = sim.cols
        let rows = sim.rows
        // Edge-to-edge: tile the full bounds and center each dot in its cell, so
        // dots span corner to corner with no dummy border gutter.
        let cellW = size.width / CGFloat(cols)
        let cellH = size.height / CGFloat(rows)
        let hasAccent = accentColor != nil

        // Crossfade spans the FULL gap to the next step (no hold), so the field
        // animates continuously. smoothstep has zero velocity at both ends, so
        // consecutive steps meet seamlessly with no visible snap.
        let window = max(0.0001, stepGap)
        let elapsed = lastStep.distance(to: now)
        let raw = min(1.0, max(0.0, elapsed / window))
        let fraction = raw * raw * (3 - 2 * raw)   // smoothstep ease

        for y in 0..<rows {
            let py = (CGFloat(y) + 0.5) * cellH
            for x in 0..<cols {
                let i = y * cols + x
                let px = (CGFloat(x) + 0.5) * cellW
                let auto = sim.value(at: i, fraction: fraction)
                let rip = rippleLevel(px: px, py: py, now: now)
                let intensity = max(auto, rip)
                // Only lit cells are drawn. Ripple cells out-shine the ambient
                // field so the user's tap reads clearly.
                let opacity = max(peakOpacity * auto, ripplePeakOpacity * rip)
                if opacity < 0.012 { continue }

                let s = dotSize * (0.82 + 0.4 * intensity)
                let rect = CGRect(x: px - s / 2, y: py - s / 2, width: s, height: s)
                let path = Path(roundedRect: rect, cornerRadius: s * 0.32, style: .continuous)

                let fill: Color
                if hasAccent, let accent = accentColor {
                    fill = baseColor.mix(with: accent, by: min(1.0, intensity * intensity * 1.2))
                } else {
                    fill = baseColor
                }
                context.fill(path, with: .color(fill.opacity(opacity)))
            }
        }
    }
}

// MARK: - Automaton

/// The stateful reaction-diffusion grid behind `PixelGlyphField`. Held in
/// `@State` as a value type — `step()` recomputes the whole grid every tick,
/// which is trivial at this resolution (~700 cells × 8 neighbors at ≤4 Hz).
private struct GlyphAutomaton {
    private(set) var cols = 0
    private(set) var rows = 0

    /// Current grid (the step just computed) and the prior grid, kept so the
    /// view can pop previous→current within a tick.
    private var current: [Double] = []
    private var previous: [Double] = []
    private var rng = SeededGenerator(seed: 0x5EED_1234)

    // --- Reaction-diffusion tuning -----------------------------------------
    // Tuned so glyphs stay SMALL and FOCUSED: low spread + high ambient decay
    // means a stamped glyph bleeds out only a cell or two before poofing away,
    // rather than growing into a screen-filling colony.
    //   survival  — how much of a cell's own value carries to the next step
    //   spread    — gain pulled in from each lit neighbor (kept low → tight)
    //   crowd     — neighbor-sum above which a cell is "saturated"…
    //   coreDecay — …and gets suppressed, hollowing filled cores → poof
    //   ambient   — flat erosion each step (high → fast dissipation)
    private let survival = 0.5
    private let spread = 0.1
    private let crowd = 2.2
    private let coreDecay = 0.32
    private let ambient = 0.05
    /// How many fresh glyphs to stamp per step (kept low so the field stays a
    /// few scattered clusters, not a busy screen).
    private let stampsPerStep = 1
    /// Extra chance for a second glyph, for a little variation in density.
    private let secondStampChance = 0.4
    /// If total energy drops below this, force-seed so it never goes dark.
    private let lowEnergy = 2.5

    mutating func configure(cols: Int, rows: Int) {
        self.cols = max(1, cols)
        self.rows = max(1, rows)
        current = Array(repeating: 0, count: self.cols * self.rows)
        previous = current
        // Seed a few colonies and pre-warm a little so the field opens with a
        // handful of small clusters already mid-bloom.
        for _ in 0..<4 { stampRandomGlyph() }
        for _ in 0..<3 { step(allowStamp: false) }
    }

    /// Eased value for cell `i`, interpolating previous→current by `fraction`.
    func value(at i: Int, fraction: Double) -> Double {
        guard i >= 0, i < current.count else { return 0 }
        let p = i < previous.count ? previous[i] : 0
        return p + (current[i] - p) * fraction
    }

    mutating func step(allowStamp: Bool = true) {
        guard cols > 0, rows > 0 else { return }
        previous = current
        var next = [Double](repeating: 0, count: current.count)

        for y in 0..<rows {
            for x in 0..<cols {
                let i = y * cols + x
                let v = current[i]

                // 8-neighbor (Moore) sum → rounder colonies than 4-neighbor.
                var n = 0.0
                for dy in -1...1 {
                    let ny = y + dy
                    if ny < 0 || ny >= rows { continue }
                    for dx in -1...1 where !(dx == 0 && dy == 0) {
                        let nx = x + dx
                        if nx < 0 || nx >= cols { continue }
                        n += current[ny * cols + nx]
                    }
                }

                var nv = v * survival + n * spread
                if n > crowd { nv *= coreDecay }   // saturated core → hollow out
                nv -= ambient
                next[i] = min(1.0, max(0.0, nv))
            }
        }
        current = next

        guard allowStamp else { return }
        for _ in 0..<stampsPerStep { stampRandomGlyph() }
        if Double.random(in: 0..<1, using: &rng) < secondStampChance {
            stampRandomGlyph()
        }
        if totalEnergy() < lowEnergy {
            stampRandomGlyph()
            stampRandomGlyph()
        }
    }

    private func totalEnergy() -> Double {
        current.reduce(0, +)
    }

    /// Stamp one randomly-chosen seed glyph at a random location, taking the max
    /// with whatever's already there (so it ignites a colony).
    private mutating func stampRandomGlyph() {
        guard cols > 1, rows > 1 else { return }
        let glyph = Self.glyphs.randomElement(using: &rng) ?? [(0, 0)]
        // Stamp anywhere across the full grid (edge to edge) — glyphs near the
        // edges are fine and wanted.
        let cx = Int.random(in: 0..<cols, using: &rng)
        let cy = Int.random(in: 0..<rows, using: &rng)
        for (dx, dy) in glyph {
            let x = cx + dx, y = cy + dy
            guard x >= 0, x < cols, y >= 0, y < rows else { continue }
            let i = y * cols + x
            current[i] = max(current[i], 1.0)
        }
    }

    /// The designed seed glyphs — small (≤3×3) so colonies stay focused. Each is
    /// a set of (dx, dy) offsets from a center cell; the automaton blooms each
    /// into a small organic cluster. Variety here keeps the shapes distinct.
    private static let glyphs: [[(Int, Int)]] = [
        // single spore
        [(0, 0)],
        // plus
        [(0, 0), (1, 0), (-1, 0), (0, 1), (0, -1)],
        // hollow ring (3×3 minus center)
        [(-1, -1), (0, -1), (1, -1), (-1, 0), (1, 0), (-1, 1), (0, 1), (1, 1)],
        // diagonal slash
        [(-1, -1), (0, 0), (1, 1)],
        // X
        [(-1, -1), (1, -1), (0, 0), (-1, 1), (1, 1)],
        // T
        [(-1, -1), (0, -1), (1, -1), (0, 0), (0, 1)],
        // 2×2 blob
        [(0, 0), (1, 0), (0, 1), (1, 1)],
        // little pair
        [(0, 0), (1, 0)],
        // L corner
        [(0, -1), (0, 0), (1, 0)],
        // diamond
        [(0, -1), (-1, 0), (1, 0), (0, 1)],
        // chevron up
        [(-1, 0), (0, -1), (1, 0)],
        // chevron down
        [(-1, 0), (0, 1), (1, 0)],
        // vertical bar
        [(0, -1), (0, 0), (0, 1)],
        // horizontal bar
        [(-1, 0), (0, 0), (1, 0)],
        // triangle
        [(0, -1), (-1, 1), (1, 1)],
        // twin verticals (gap)
        [(-1, -1), (-1, 0), (-1, 1), (1, -1), (1, 0), (1, 1)],
        // staircase
        [(-1, 1), (0, 0), (1, -1), (0, 1)],
        // spaced trio
        [(-2, 0), (0, 0), (2, 0)],
        // arrowhead
        [(0, -1), (-1, 0), (1, 0), (0, 0), (0, 1)],
        // scattered cluster
        [(-1, -1), (1, 0), (0, 1), (2, 1)],
    ]
}

/// Deterministic PRNG so the field is reproducible and `Sendable`-clean.
/// xorshift64* — fast, decent distribution for visual seeding.
private struct SeededGenerator: RandomNumberGenerator {
    private var state: UInt64

    init(seed: UInt64) {
        state = seed != 0 ? seed : 0x9E37_79B9_7F4A_7C15
    }

    mutating func next() -> UInt64 {
        state ^= state >> 12
        state ^= state << 25
        state ^= state >> 27
        return state &* 0x2545_F491_4F6C_DD1D
    }
}

#Preview("Glyph field") {
    ZStack {
        Color.black.ignoresSafeArea()
        PixelGlyphField()
            .ignoresSafeArea()
    }
    .preferredColorScheme(.dark)
}

File: wagerproof-android-native/core/design/src/main/kotlin/com/wagerproof/core/design/backgrounds/PixelGlyphField.kt
package com.wagerproof.core.design.backgrounds

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.lerp
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onSizeChanged
import com.wagerproof.core.design.tokens.AppColors
import kotlin.math.exp
import kotlin.math.roundToInt
import kotlinx.coroutines.delay

/**
 * Command channel for spawning tap-style ripples in a [PixelGlyphField] from
 * OUTSIDE the field — port of iOS `GlyphRippleEmitter` (GlyphRippleOnChange.swift).
 * E.g. tapping the agent avatar ripples the background pixels behind it.
 *
 * [Pulse.point] is in the field's OWN coordinate space. A screen-anchored field
 * (`PixelWaveBackground(screenAnchored = true)`) draws at window (0,0) sized to
 * the full screen, so a WINDOW-space point maps 1:1 — pass the tapped view's
 * window-space center (`onGloballyPositioned` → `boundsInWindow().center`).
 */
class GlyphRippleEmitter {
    data class Pulse(
        val point: Offset,
        /**
         * Monotonic id so two ripples at the same point still read as distinct
         * events (a plain point wouldn't restart an effect on a repeat tap).
         */
        val token: Int,
    )

    var pulse: Pulse by mutableStateOf(Pulse(Offset.Zero, 0))
        private set

    /** Spawn a ripple centered on [at] (field-local / screen-anchored-window). */
    fun emit(at: Offset) {
        pulse = Pulse(at, pulse.token + 1)
    }
}

/**
 * Compose analog of iOS's `\.glyphRippleEmitter` environment entry: a container
 * hosting a pixelwave background injects its emitter so descendants can ripple
 * it without coupling to the background composable.
 */
val LocalGlyphRippleEmitter = staticCompositionLocalOf<GlyphRippleEmitter?> { null }

/**
 * A stepped, organic field of small pixel "glyphs" that bloom and dissipate
 * like bacteria colonies — port of iOS `PixelGlyphField.swift`.
 *
 * Unlike [PixelDotBackground] (a stateless pure-function-of-time field), a
 * spreading/decaying colony needs state that evolves in discrete steps: a
 * small reaction-diffusion cellular automaton runs over the dot grid. Designed
 * seed glyphs are periodically stamped on; the automaton grows them a little,
 * hollows saturated cores, and erodes them with ambient decay — each glyph
 * stays small and focused, then poofs away.
 *
 * The display eases previous→current across the full gap to the next step, so
 * motion is continuous — no hold between refreshes. Only lit cells are drawn.
 */
@Composable
fun PixelGlyphField(
    modifier: Modifier = Modifier,
    /** Step cadence, cycled in order. One value = steady beat. */
    intervals: List<Double> = listOf(0.15),
    baseColor: Color = Color.White,
    accentColor: Color? = AppColors.appPrimary,
    spacing: Float = 26f,
    dotSize: Float = 5.5f,
    peakOpacity: Double = 0.45,
    /**
     * Seeds the automaton's PRNG — per-card seeds keep multiple fields out of
     * lockstep. Default matches the shared iOS hero constant.
     */
    seed: Long = 0x5EED_1234L,
    /** External ripple trigger; null = ripples only on direct taps. */
    rippleEmitter: GlyphRippleEmitter? = null,
    /** Disable the field's own tap-ripple gesture (screen-anchored use). */
    tapRipples: Boolean = true,
    reduceMotion: Boolean = false,
) {
    val cadence = remember(intervals) { if (intervals.isEmpty()) listOf(0.3) else intervals }
    val sim = remember { GlyphAutomaton() }
    var fieldSize by remember { mutableStateOf(Size.Zero) }
    // Bumped every automaton step so the (otherwise time-only) canvas redraw
    // picks up the new grid.
    var stepTick by remember { mutableIntStateOf(0) }
    var lastStep by remember { mutableStateOf(DesignClock.nowSeconds()) }
    // Crossfade window = time until the NEXT step, so the ease fills the whole
    // gap and the field never sits still.
    var stepGap by remember { mutableStateOf(cadence.first()) }
    val ripples = remember { mutableStateOf(listOf<GlyphRipple>()) }

    val time by rememberFrameSeconds(paused = reduceMotion)

    fun addRipple(at: Offset) {
        if (reduceMotion) return
        val now = DesignClock.nowSeconds()
        var active = ripples.value.filter { now - it.start <= RIPPLE_DURATION }
        active = active + GlyphRipple(at, now)
        if (active.size > 8) active = active.takeLast(8)
        ripples.value = active
    }

    // Stepping loop; restarts if reduce-motion toggles. Frozen → the field
    // holds its last state (still drawn — just static).
    LaunchedEffect(reduceMotion, cadence) {
        if (reduceMotion) return@LaunchedEffect
        var idx = 0
        while (true) {
            delay((cadence[idx] * 1000).toLong())
            sim.step()
            lastStep = DesignClock.nowSeconds()
            stepTick++
            idx = (idx + 1) % cadence.size
            stepGap = cadence[idx]
        }
    }

    // External ripples bypass hit testing, so they fire even when the field is
    // inert behind a scroll. token 0 = the initial no-op pulse.
    val pulse = rippleEmitter?.pulse
    LaunchedEffect(pulse) {
        if (pulse != null && pulse.token > 0) addRipple(pulse.point)
    }

    val tapModifier = if (tapRipples) {
        Modifier.pointerInput(reduceMotion) {
            // The field sits behind foreground content, so buttons still get
            // their taps first; only taps that fall through ripple.
            detectTapGestures { addRipple(it) }
        }
    } else {
        Modifier
    }

    Canvas(
        modifier = modifier
            .onSizeChanged { newSize ->
                val cols = maxOf(2, (newSize.width / spacing).roundToInt())
                val rows = maxOf(2, (newSize.height / spacing).roundToInt())
                sim.configure(cols, rows, seed)
                lastStep = DesignClock.nowSeconds()
                stepTick++
                fieldSize = Size(newSize.width.toFloat(), newSize.height.toFloat())
            }
            .then(tapModifier),
    ) {
        stepTick // read so step invalidates the draw even between frame ticks
        val now = time
        if (size.width <= 0f || size.height <= 0f || sim.cols <= 0 || sim.rows <= 0) return@Canvas

        val cols = sim.cols
        val rows = sim.rows
        // Edge-to-edge: tile the full bounds and center each dot in its cell.
        val cellW = size.width / cols
        val cellH = size.height / rows

        // smoothstep ease across the full inter-step gap — zero velocity at
        // both ends, so consecutive steps meet seamlessly with no snap.
        val window = maxOf(0.0001, stepGap)
        val raw = ((now - lastStep) / window).coerceIn(0.0, 1.0)
        val fraction = raw * raw * (3 - 2 * raw)

        val activeRipples = ripples.value.filter { now - it.start in 0.0..RIPPLE_DURATION }

        for (y in 0 until rows) {
            val py = (y + 0.5f) * cellH
            for (x in 0 until cols) {
                val i = y * cols + x
                val px = (x + 0.5f) * cellW
                val auto = sim.value(i, fraction)
                val rip = rippleLevel(activeRipples, px, py, now)
                val intensity = maxOf(auto, rip)
                // Ripple cells out-shine the ambient field so a tap reads clearly.
                val opacity = maxOf(peakOpacity * auto, RIPPLE_PEAK_OPACITY * rip)
                if (opacity < 0.012) continue

                val s = dotSize * (0.82f + 0.4f * intensity.toFloat())
                val fill = if (accentColor != null) {
                    lerp(baseColor, accentColor, minOf(1.0, intensity * intensity * 1.2).toFloat())
                } else {
                    baseColor
                }
                drawRoundRect(
                    color = fill.copy(alpha = opacity.toFloat()),
                    topLeft = Offset(px - s / 2f, py - s / 2f),
                    size = Size(s, s),
                    cornerRadius = CornerRadius(s * 0.32f),
                )
            }
        }
    }
}

// Tap-ripple tuning (iOS constants).
private const val RIPPLE_DURATION = 1.5 // seconds
private const val RIPPLE_SPEED = 430.0 // ring expansion, px/s
private const val RIPPLE_RING_WIDTH = 26.0 // ring softness, px
private const val RIPPLE_PEAK_OPACITY = 0.72

private class GlyphRipple(val origin: Offset, val start: Double)

/**
 * Combined brightness (0..1) from all active ripples for a cell at (px, py):
 * each ripple is a soft expanding gaussian ring that fades with age.
 */
private fun rippleLevel(ripples: List<GlyphRipple>, px: Float, py: Float, now: Double): Double {
    if (ripples.isEmpty()) return 0.0
    var level = 0.0
    for (ripple in ripples) {
        val age = now - ripple.start
        if (age < 0 || age > RIPPLE_DURATION) continue
        val radius = age * RIPPLE_SPEED
        val dx = (px - ripple.origin.x).toDouble()
        val dy = (py - ripple.origin.y).toDouble()
        val delta = Math.sqrt(dx * dx + dy * dy) - radius
        val ring = exp(-(delta * delta) / (2 * RIPPLE_RING_WIDTH * RIPPLE_RING_WIDTH))
        val fade = 1 - age / RIPPLE_DURATION
        level = maxOf(level, ring * fade)
    }
    return level
}

// MARK: - Automaton

/**
 * The stateful reaction-diffusion grid behind [PixelGlyphField]. `step()`
 * recomputes the whole grid every tick — trivial at ~700 cells × 8 neighbors
 * at ≤4 Hz. Tuning keeps glyphs SMALL and FOCUSED: low spread + high ambient
 * decay means a stamped glyph bleeds out a cell or two before poofing away.
 */
internal class GlyphAutomaton {
    var cols = 0
        private set
    var rows = 0
        private set

    private var current = DoubleArray(0)
    private var previous = DoubleArray(0)
    private var rng = SeededRandom(0x5EED_1234L)

    // survival — cell's own carry; spread — gain per lit neighbor;
    // crowd/coreDecay — saturation hollowing; ambient — flat erosion.
    private val survival = 0.5
    private val spread = 0.1
    private val crowd = 2.2
    private val coreDecay = 0.32
    private val ambient = 0.05
    private val stampsPerStep = 1
    private val secondStampChance = 0.4

    /** If total energy drops below this, force-seed so it never goes dark. */
    private val lowEnergy = 2.5

    fun configure(cols: Int, rows: Int, seed: Long = 0x5EED_1234L) {
        this.cols = maxOf(1, cols)
        this.rows = maxOf(1, rows)
        // Re-seed before stamping so pre-warmed colonies (and every step after)
        // differ per field — distinct cards never bloom in lockstep.
        rng = SeededRandom(seed)
        current = DoubleArray(this.cols * this.rows)
        previous = current.copyOf()
        // Seed a few colonies and pre-warm so the field opens mid-bloom.
        repeat(4) { stampRandomGlyph() }
        repeat(3) { step(allowStamp = false) }
    }

    /** Eased value for cell [i], interpolating previous→current by [fraction]. */
    fun value(i: Int, fraction: Double): Double {
        if (i < 0 || i >= current.size) return 0.0
        val p = if (i < previous.size) previous[i] else 0.0
        return p + (current[i] - p) * fraction
    }

    fun step(allowStamp: Boolean = true) {
        if (cols <= 0 || rows <= 0) return
        previous = current
        val next = DoubleArray(current.size)

        for (y in 0 until rows) {
            for (x in 0 until cols) {
                val i = y * cols + x
                val v = current[i]

                // 8-neighbor (Moore) sum → rounder colonies than 4-neighbor.
                var n = 0.0
                for (dy in -1..1) {
                    val ny = y + dy
                    if (ny < 0 || ny >= rows) continue
                    for (dx in -1..1) {
                        if (dx == 0 && dy == 0) continue
                        val nx = x + dx
                        if (nx < 0 || nx >= cols) continue
                        n += current[ny * cols + nx]
                    }
                }

                var nv = v * survival + n * spread
                if (n > crowd) nv *= coreDecay // saturated core → hollow out
                nv -= ambient
                next[i] = nv.coerceIn(0.0, 1.0)
            }
        }
        current = next

        if (!allowStamp) return
        repeat(stampsPerStep) { stampRandomGlyph() }
        if (rng.nextDouble() < secondStampChance) stampRandomGlyph()
        if (current.sum() < lowEnergy) {
            stampRandomGlyph()
            stampRandomGlyph()
        }
    }

    /** Stamp one random seed glyph at a random location, max-merged in. */
    private fun stampRandomGlyph() {
        if (cols <= 1 || rows <= 1) return
        val glyph = GLYPHS[rng.nextInt(GLYPHS.size)]
        val cx = rng.nextInt(cols)
        val cy = rng.nextInt(rows)
        for ((dx, dy) in glyph) {
            val x = cx + dx
            val y = cy + dy
            if (x < 0 || x >= cols || y < 0 || y >= rows) continue
            val i = y * cols + x
            current[i] = maxOf(current[i], 1.0)
        }
    }

    companion object {
        /**
         * The 20 designed seed glyphs — small (≤3×3) (dx, dy) offset sets so
         * colonies stay focused. Copied verbatim from iOS.
         */
        private val GLYPHS: List<List<Pair<Int, Int>>> = listOf(
            // single spore
            listOf(0 to 0),
            // plus
            listOf(0 to 0, 1 to 0, -1 to 0, 0 to 1, 0 to -1),
            // hollow ring (3×3 minus center)
            listOf(-1 to -1, 0 to -1, 1 to -1, -1 to 0, 1 to 0, -1 to 1, 0 to 1, 1 to 1),
            // diagonal slash
            listOf(-1 to -1, 0 to 0, 1 to 1),
            // X
            listOf(-1 to -1, 1 to -1, 0 to 0, -1 to 1, 1 to 1),
            // T
            listOf(-1 to -1, 0 to -1, 1 to -1, 0 to 0, 0 to 1),
            // 2×2 blob
            listOf(0 to 0, 1 to 0, 0 to 1, 1 to 1),
            // little pair
            listOf(0 to 0, 1 to 0),
            // L corner
            listOf(0 to -1, 0 to 0, 1 to 0),
            // diamond
            listOf(0 to -1, -1 to 0, 1 to 0, 0 to 1),
            // chevron up
            listOf(-1 to 0, 0 to -1, 1 to 0),
            // chevron down
            listOf(-1 to 0, 0 to 1, 1 to 0),
            // vertical bar
            listOf(0 to -1, 0 to 0, 0 to 1),
            // horizontal bar
            listOf(-1 to 0, 0 to 0, 1 to 0),
            // triangle
            listOf(0 to -1, -1 to 1, 1 to 1),
            // twin verticals (gap)
            listOf(-1 to -1, -1 to 0, -1 to 1, 1 to -1, 1 to 0, 1 to 1),
            // staircase
            listOf(-1 to 1, 0 to 0, 1 to -1, 0 to 1),
            // spaced trio
            listOf(-2 to 0, 0 to 0, 2 to 0),
            // arrowhead
            listOf(0 to -1, -1 to 0, 1 to 0, 0 to 0, 0 to 1),
            // scattered cluster
            listOf(-1 to -1, 1 to 0, 0 to 1, 2 to 1),
        )
    }
}

/**
 * Deterministic xorshift64* PRNG — same algorithm/seed handling as the iOS
 * `SeededGenerator` so per-seed glyph patterns stay reproducible.
 */
internal class SeededRandom(seed: Long) {
    private var state: Long = if (seed != 0L) seed else 0x9E37_79B9_7F4A_7C15UL.toLong()

    private fun next(): Long {
        state = state xor (state ushr 12)
        state = state xor (state shl 25)
        state = state xor (state ushr 27)
        return state * 0x2545_F491_4F6C_DD1DUL.toLong()
    }

    /** Uniform in `0 until bound`. */
    fun nextInt(bound: Int): Int {
        require(bound > 0)
        val r = next().toULong()
        return (r % bound.toULong()).toInt()
    }

    /** Uniform in `0.0 until 1.0` (top 53 bits, standard construction). */
    fun nextDouble(): Double = (next().toULong() shr 11).toDouble() / (1L shl 53).toDouble()
}
